onnxruntime/core/providers/cpu/nn/pool_base.h

```diff
@@ -183,37 +199,56 @@ class PoolBase {
                                 const int64_t kernel,
                                 int64_t* pad_head,
                                 int64_t* pad_tail,
+                                int64_t dilation,
+                                int64_t ceil_mode,
                                 int64_t* out_size) const {
     if (auto_pad_ != AutoPadType::NOTSET) {
       switch (auto_pad_) {
         case AutoPadType::VALID:
           *pad_head = 0;
           *pad_tail = 0;
-          *out_size = (in_size - kernel) / stride + 1;
+          if (ceil_mode == 0) {
+            *out_size = (in_size - dilation * (kernel - 1) - 1) / stride + 1;
+          } else {
+            *out_size = (int64_t)ceil((in_size - dilation * (kernel - 1) - 1) / (float) stride + 1);
+          }
           break;
         case AutoPadType::SAME_LOWER: {
           int64_t legacy_target_size = (in_size + stride - 1) / stride;
           int64_t pad_needed = (legacy_target_size - 1) * stride + kernel - in_size;
           *pad_head = (pad_needed + 1) / 2;
           *pad_tail = pad_needed - *pad_head;
-          *out_size = (in_size + pad_needed - kernel) / stride + 1;
+          if (ceil_mode == 0) {
+            *out_size = (in_size + pad_needed - dilation * (kernel - 1) - 1) / stride + 1;
+          } else {
+            *out_size = (int64_t)ceil((in_size - dilation * (kernel - 1) - 1) / (float) stride + 1);
+          }
           break;
         }
         case AutoPadType::SAME_UPPER: {
           int64_t legacy_target_size = (in_size + stride - 1) / stride;
           int64_t pad_needed = (legacy_target_size - 1) * stride + kernel - in_size;
           *pad_head = pad_needed / 2;
           *pad_tail = pad_needed - *pad_head;
-          *out_size = (in_size + pad_needed - kernel) / stride + 1;
+          if (ceil_mode == 0) {
+            *out_size = (in_size + pad_needed - dilation * (kernel - 1) - 1) / stride + 1;
+          } else {
+            *out_size = (int64_t)ceil((in_size - dilation * (kernel - 1) - 1) / (float) stride + 1);
+          }
           break;
         }
         default: {
           ORT_THROW("Unsupported AutoPad Type.");
         }
       }
     } else {
-      *out_size = static_cast<int64_t>(
-          static_cast<float>(in_size + *pad_head + *pad_tail - kernel) / stride + 1);
+      if (ceil_mode == 0) {
+        *out_size = static_cast<int64_t>(
+            static_cast<float>(in_size + *pad_head + *pad_tail - dilation * (kernel - 1) -1) / (float) stride + 1);
```

Comment:
>in_size + *pad_head + *pad_tail - dilation * (kernel - 1 [](start = 31, length = 56)

This code is repeated in many places. @Pranav Sharma do you think it might be benefitial to wrap it up in  an inline function with a clear name?

Reply:
sure, sounds good.

Reply:
refactored into separate function to make it modular.